server/routes/visit/visitUtils.test.ts

```diff
@@ -36,4 +36,91 @@ describe('Visit utils', () => {
+          update: false,
+          cancel: false,
+          clearNotifications: false,
+        })
+      })
+    })
+
+    describe('BOOKED visit', () => {
+      describe('update action', () => {
+        it('should set update to false if visit has started', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-01T09:00:00') })
+          expect(getAvailableVisitActions(params).update).toBe(false)
+        })
+
+        it('should set update to true before visit has started', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-01T08:59:59') })
+          expect(getAvailableVisitActions(params).update).toBe(true)
+        })
+
+        it('should set update to false if before start time but prisoner has been released', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-01T08:59:59') })
+          params.notifications = [{ type: 'PRISONER_RELEASED_EVENT' }] as VisitBookingDetailsDto['notifications']
+          expect(getAvailableVisitActions(params).update).toBe(false)
+        })
+
+        it('should set update to false if before start time but prisoner has been transferred', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-01T08:59:59') })
+          params.notifications = [{ type: 'PRISONER_RECEIVED_EVENT' }] as VisitBookingDetailsDto['notifications']
+          expect(getAvailableVisitActions(params).update).toBe(false)
+        })
+      })
+
+      describe('cancel action', () => {
+        it('should set cancel to true before visit start time', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-01T08:59:59') })
+          expect(getAvailableVisitActions(params).cancel).toBe(true)
+        })
+
+        it('should set cancel to true after visit start time, up to CANCELLATION_LIMIT_DAYS (28)', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-29T08:59:59') }) // startTimestamp + 28 days
+          expect(getAvailableVisitActions(params).cancel).toBe(true)
+        })
+
+        it('should set cancel to false after CANCELLATION_LIMIT_DAYS (28)', () => {
+          jest.useFakeTimers({ now: new Date('2025-04-29T09:00:00') }) // startTimestamp + 28 days + 1 second
+          expect(getAvailableVisitActions(params).cancel).toBe(false)
+        })
+      })
+
+      describe('clearNotifications action', () => {
+        it('should set clearNotifications to false if no visit notifications', () => {
+          expect(getAvailableVisitActions(params).clearNotifications).toBe(false)
+        })
+
+        it('should set clearNotifications to true if a visit notification is set', () => {
+          params.notifications = [{ type: 'NON_ASSOCIATION_EVENT' }] as VisitBookingDetailsDto['notifications']
+          expect(getAvailableVisitActions(params).clearNotifications).toBe(true)
+        })
+
+        it('should set clearNotifications to false if visit notifications includes PRISON_VISITS_BLOCKED_FOR_DATE', () => {
```

Comment:
This is a change to previous behaviour: notifications cannot now be cleared for a blocked date, even if there are other notifications set (because the visit cannot go ahead so must be updated or cancelled).